### PR TITLE
ENT-5134 Discoverable Hibernate Session Factory Factory

### PR DESCRIPTION
**Pull request by blsemo**
Removing the database specific if/else statements from HibernateConfiguration and replace them with SessionFactoryFactory implementations. These are discovered via the service loader and can each handle one type of jdbc connection (i.e. database type). Thus the configuration becomes linear set of instructions for that specific database flavour.
This also allows to add HibernateConfiguration for other database types without modifying the actual HibernateConfiguration class, and allows us to use the Open Source version as Open Core library for the Enterprise offering.

Tested by running the bank-of-corda sample with H2 locally.

### CODE REVIEW

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/HibernateConfiguration.kt [hunk 1]
```diff
@@ -5,60 +5,47 @@ import net.corda.core.internal.NamedCacheFactory
...
-                // Create a custom type that will map a blob to byteA in postgres and as a normal blob for all other dbms.
-                // This is required for the Checkpoints as a workaround for the issue that postgres has on azure.
-                if (jdbcUrl.contains(":postgresql:", ignoreCase = true)) {
-                    applyBasicType(MapBlobToPostgresByteA, MapBlobToPostgresByteA.name)
-                } else {
-                    applyBasicType(MapBlobToNormalBlob, MapBlobToNormalBlob.name)
-                }
-
-                return build()
+        for( sff in serviceLoader.iterator()){
```
Comment (mnesbit) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/HibernateConfiguration.kt]: formatting looks a little ragged

Comment (rick-r3) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/HibernateConfiguration.kt]: And you don't need the `.iterator()`.  It's `Iterable` so Kotlin understands.

**Comment by blsemo** [comment]
run tests: smoke

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/HibernateConfiguration.kt [hunk 2]
```diff
@@ -5,60 +5,55 @@ import net.corda.core.internal.NamedCacheFactory
...
-                return build()
+            val presentFactories = mutableListOf<String>()
+
+            for (sff in serviceLoader.iterator()) {
+                if (sff.canHandleDatabase(jdbcUrl)) {
+                    return sff
+                }
+                presentFactories.add(sff.databaseType)
             }
+            throw HibernateConfigException("Failed to find a SessionFactoryFactory to handle $jdbcUrl - factories present for [${presentFactories.joinToString { ", " }}]")
```
Comment (szymonsztuka) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/HibernateConfiguration.kt]: You don't need to build manually the list of dbnames as you will end up with exactly the same content as the original list from ServiceLoader.
`sff` - not Java naming convetion
Redundant  `joinTostring` - Java does it automatically by calling toString on a list 
The code doesn't cater about unusual case when you have duplicated definitions  (two classes implementing handing the same jdbc connection string), should we at least warn in logs? 
My overall suggestion would be (also more Kotlin style without explicit loops):
```
            val sessionFactories = serviceLoader.filter { it.canHandleDatabase(jdbcUrl) }
            when (sessionFactories.size) {
                0 -> throw HibernateConfigException("Failed to find a SessionFactoryFactory to handle $jdbcUrl " +
                        "- factories present for ${ serviceLoader.map { it.databaseType }}]")
                1 -> return sessionFactories.single()
                else -> throw HibernateConfigException("Find several SessionFactoryFactory classes to handle $jdbcUrl " +
                        "- classes ${ sessionFactories.map { it.javaClass.canonicalName }}")
            }
```

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+import javax.persistence.AttributeConverter
+
+abstract class BaseSessionFactoryFactory : CordaSessionFactoryFactory {
+    companion object{
+        private val logger = contextLogger()
+    }
+
+    open fun buildHibernateConfig(databaseConfig: DatabaseConfig, metadataSources: MetadataSources): Configuration {
+        val hbm2dll: String =
+                if(databaseConfig.initialiseSchema && databaseConfig.initialiseAppSchema == SchemaInitializationType.UPDATE) {
```
Comment (szymonsztuka) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt]: There may be differences between OS and ENT how `initialiseAppSchema` and `hibernate.hbm2ddl.auto` are handled for H2 database.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            // to avoid OOM when large blobs might get logged.
+            applyBasicType(CordaMaterializedBlobType, CordaMaterializedBlobType.name)
+            applyBasicType(CordaWrapperBinaryType, CordaWrapperBinaryType.name)
+            applyBasicType(MapBlobToNormalBlob, MapBlobToNormalBlob.name)
+
+            return build()
+        }
+    }
+
+    open fun buildSessionFactory(
```
Comment (szymonsztuka) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt]: Should this be final as it's not database vendor specific?

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/CordaSessionFactoryFactory.kt [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+interface CordaSessionFactoryFactory {
+    val databaseType: String
+    fun canHandleDatabase(jdbcUrl: String): Boolean
+    fun makeSessionFactoryForSchemas(
+            databaseConfig: DatabaseConfig,
+            schemas: Set<MappedSchema>,
+            customClassLoader: ClassLoader?,
+            attributeConverters: Collection<AttributeConverter<*, *>>): SessionFactory
+    fun getExtraConfiguration(key: String): Any?
```
Comment (szymonsztuka) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/CordaSessionFactoryFactory.kt:18]: Will it be in use? If not maybe let's not create an additional method for future use which will be never used.
A factory class should be  self contained in principle and abstracting away any internals to the client code (caller), so I'm not sure why the caller would like to know the configuration, which was used to create the factory?

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/PostgresSessionFactoryFactory.kt [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            applyBasicType(MapBlobToPostgresByteA, MapBlobToPostgresByteA.name)
+
+            return build()
+        }
+    }
+
+    override fun canHandleDatabase(jdbcUrl: String): Boolean = jdbcUrl.contains(":postgresql:")
+
+    // Maps to a byte array on postgres.
+    object MapBlobToPostgresByteA : AbstractSingleColumnStandardBasicType<ByteArray>(VarbinaryTypeDescriptor.INSTANCE, PrimitiveByteArrayTypeDescriptor.INSTANCE) {
```
Comment (szymonsztuka) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/PostgresSessionFactoryFactory.kt:30]: Maybe split the long line to please Detekt?

**Review by szymonsztuka — COMMENTED** [review]
I think the approach with abstract base class is not optimal here as:
we want to parameterise `HibernateSessionFactory` creation depending on database for only two  steps:
1) when creating Hibernate sessionFactory we want to add custom configuration properties
`` ('config' in 
    val sessionFactory = buildSessionFactory(config, metadataSources, customClassLoader))``

2) when creating `HibernateMetadata` we want to apply specific type mappings for Hibernate `@Type` annotation for our Corda specific types ( 2b) however we need always apply to the same mappings but just decide which implementation to use). See:

```
fun buildHibernateMetadata(metadataBuilder: MetadataBuilder, jdbcUrl:String, attributeConverters: Collection<AttributeConverter<*, *>>): Metadata {
            metadataBuilder
            applySpecificTypes // <-- the type implementaion depends on a database vendor
            )
```

The better approach would be to create a template method (pattern), the whole code would be in that class which is final and we can parameterised extension points for it (passed via e.g. constructor),e.g:    

```
//instead of *SesionFactoryFactory: 
 interface CordaHibernateOptions {
    val properites :  Properties  //for session config
    val types : List<Any!>   //object classes for MetadataBuilder “applySpecificTypes”
 } 

 val databaseHibernateOption = ServiceLocator.getHibernateOptions(jdbc)
 val x = HibernateSessionBuilder(databseHibernateOption.properties, databseHibernateOption.types 
```

 In such way we avoid some code duplication when overriding the whole methods (and potentially introducing more bugs), we impose specifically what needs to be parameterised.
Even stronger typed version would be:
```
class CordaHibernateOptions {
    val properites :  Properties  //for session config
    val cordaBlobType : 
   val cordaWrapperBinaryType: 
} 
```
As in Corda we have maximum 2-4  custom types. Our custom converters are database agnostic though.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt (hunk 3 above)
Comment (blsemo) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt]: Yes, but we want to get rid of those.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/CordaSessionFactoryFactory.kt (hunk 5 above)
Comment (blsemo) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/CordaSessionFactoryFactory.kt:18]: Yes, this one is used in Enterprise to hand out extra config parameters to functions in the vault (in this case vault query batch size for SqlServer only)

**Comment by blsemo** [comment]
Unfortunately, in the enterprise version we do some more "interesting" things in the hibernate config (e.g. handle nationalised strings differently for some dbs, encapsulate schema names in quotes in post-gres etc.) A list of injectable configs would make this quite tricky.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    override fun buildHibernateMetadata(metadataBuilder: MetadataBuilder, attributeConverters: Collection<AttributeConverter<*, *>>): Metadata {
+        metadataBuilder.run {
+            attributeConverters.forEach { applyAttributeConverter(it) }
+            // Register a tweaked version of `org.hibernate.type.MaterializedBlobType` that truncates logged messages.
+            // to avoid OOM when large blobs might get logged.
+            applyBasicType(CordaMaterializedBlobType, CordaMaterializedBlobType.name)
+            applyBasicType(CordaWrapperBinaryType, CordaWrapperBinaryType.name)
+            applyBasicType(MapBlobToNormalBlob, MapBlobToNormalBlob.name)
+
+            return build()
```
Comment (szymonsztuka) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt]: Really oppinated - `return` could be moved up at  `metadataBuilder.run {`. level (`return metadataBuilder.run {`) . this style is used in the following method `buildSessionFactory`.  The same comments for other occurences of `buildHibernateMetadata` method.

**Review by szymonsztuka — DISMISSED** [review]
LGTM, one minor optional  style suggestion

**Comment by szymonsztuka** [comment]
> Unfortunately, in the enterprise version we do some more "interesting" things in the hibernate config (e.g. handle nationalised strings differently for some dbs, encapsulate schema names in quotes in post-gres etc.) A list of injectable configs would make this quite tricky.

Ok, it makes sense, the otehr approach could be the visitor pattern, hoevwer for the code wich changes very rarely it might be over engineering.

**Review by szymonsztuka — DISMISSED** [review]

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/HibernateConfiguration.kt [hunk 8]
```diff
@@ -5,60 +5,54 @@ import net.corda.core.internal.NamedCacheFactory
...
+                ServiceLoader.load(CordaSessionFactoryFactory::class.java, customClassLoader)
+            else
+                ServiceLoader.load(CordaSessionFactoryFactory::class.java)
+
+            val sessionFactories = serviceLoader.filter { it.canHandleDatabase(jdbcUrl) }
+            when (sessionFactories.size) {
+                0 -> throw HibernateConfigException("Failed to find a SessionFactoryFactory to handle $jdbcUrl " +
+                        "- factories present for ${serviceLoader.map { it.databaseType }}")
+                1 -> return sessionFactories.single()
+                else -> throw HibernateConfigException("Find several SessionFactoryFactory classes to handle $jdbcUrl " +
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/HibernateConfiguration.kt]: Find -> Found

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        // We set a connection provider as the auto schema generation requires it.  The auto schema generation will not
+        // necessarily remain and would likely be replaced by something like Liquibase.  For now it is very convenient though.
+        val config = Configuration(metadataSources).setProperty("hibernate.connection.provider_class", HibernateConfiguration.NodeDatabaseConnectionProvider::class.java.name)
+                .setProperty("hibernate.format_sql", "true")
+                .setProperty("javax.persistence.validation.mode", "none")
+                .setProperty("hibernate.connection.isolation", databaseConfig.transactionIsolationLevel.jdbcValue.toString())
+
+        config.setProperty("hibernate.hbm2ddl.auto", "validate")
+
+        return config
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt]: Nit-pick (but while we're at it): We could tidy this method up to use chaining of `setProperty` for `"hibernate.hbm2ddl.auto"` too and then no need for the `val` either ?

**Review by rick-r3 — COMMENTED** [review]
Couple of minor things.

**Review by rick-r3 — APPROVED** [review]

**Pull request merged**